### PR TITLE
Add boundary conditions and standardise field and timed data storage

### LINKED ISSUE #185: Add ability to add boundary conditions to object store
As well as Emissions and Footprints, for ancillary data we should also create a BoundaryConditions class or some other way to add this type of data to the object store.

This will involve data which defines mole fractions for multiple vertical boundaries (curtains) surrounding a pre-defined area (domain).

### PR DESCRIPTION
**Pull request by rt17603**
Adding new `BoundaryConditions` class to allow boundary conditions data (vertical curtains around a lat-lon domain) to be added to the object store. Tests have been added for this (based on those present within "test_emissions.py")

As a part of this the following additional changes have been made:
 - Within "_datasource.py" the `Datasource.add_timeseries` and `Datasource.add_field_data` (and derived) methods have been merged into one `Datasource.add_timed_data` method which is currently called when "time" is a coordinate within our data.
    - This method now adds the extent period (e.g. sampling_period or time_period) to the end_date to make sure the label covers the time *represented* by the data.
    - Note that this will now alter *all* daterange_str values for timeseries and fields data (the tests have been updated to reflect this). The only case where this will not change is if the "sampling_period" is unknown.
 - To code to determine the time period of the data has now been standardised between `BoundaryConditions`, `Emissions` and `Footprints` classes (split out into a separate `_infer_time.py` module)

Closes #185

### CODE REVIEW

**Comment by gareth-j** [comment]
This look great! I might leave it til Monday to start going over it, thanks Friday brain.

File: openghg/store/_footprints.py [hunk 1]
```diff
@@ -156,34 +178,34 @@ def read_file(
...
-    #         site: Site code
-    #         domain: Domain
-    #         model: Model name
-    #         height: Height
-    #         uuid: UUID of Datasource
-    #     Returns:
-    #         None
-    #     """
-    #     self._datasource_table[site][domain][model][height] = uuid
+    def lookup_uuid(self, site: str, domain: str, model: str, height: str) -> Union[str, bool]:
```
Comment (gareth-j) [inline openghg/store/_footprints.py:181]: This can be removed

File: openghg/store/_footprints.py [hunk 2]
```diff
@@ -156,34 +178,34 @@ def read_file(
...
+            model: Model name
+            height: Height
+        Returns:
+            str or dict: UUID or False if no entry
+        """
+        uuid = self._datasource_table[site][domain][model][height]
+
+        return uuid if uuid else False
+
+    def set_uuid(self, site: str, domain: str, model: str, height: str, uuid: str) -> None:
```
Comment (gareth-j) [inline openghg/store/_footprints.py:196]: This can be removed

File: openghg/store/_infer_time.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        filepath: Full path to original netcdf file
+        period: Period of measurements.
+                If specified, should be one of:
+                    - "yearly", "monthly"
+                    - suitable pandas Offset Alias
+                    - tuple of (value, unit) as would be passed to pandas.Timedelta function
+        continuous: Whether time stamps have to be continuous.
+
+    Returns:
+        str, str, str:
```
Comment (gareth-j) [inline openghg/store/_infer_time.py:39]: The type hint and return value here are different.

File: tests/store/test_boundary_conditions.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    for key in expected_metadata.keys():
+        assert metadata[key] == expected_metadata[key]
+
+
+# TODO: Add test for co2 data - need to create TEST region to match other data for this
+# TODO: Add test for multiple values within a file - continuous (maybe monthly)
+# TODO: Add test around non-continuous data and key word?
+
+
+def test_set_lookup_uuids():
```
Comment (gareth-j) [inline tests/store/test_boundary_conditions.py:130]: This can be removed

**Review by gareth-j — CHANGES_REQUESTED** [review]
Looks great! Just a few small changes and it should be ready to merge.

File: openghg/store/_boundary_conditions.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        bc_store._file_hashes[file_hash] = filepath.name
+
+        bc_store.save()
+
+        metastore.close()
+
+        return datasource_uuids
+
+    @staticmethod
+    def format() -> dict:
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/_boundary_conditions.py:178]: Ok - I'll update this to Dict[str, Any]

File: openghg/store/_boundary_conditions.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        data_format = {"dims": dims, "data_vars": data_vars, "data_types": data_types}
+
+        return data_format
+
+    def check_format(self) -> None:
+        # TODO: Create check_format() function to define and align format to
+        # expected values within database
+        pass
+
+    def lookup_uuid(self, species: str, bc_input: str, domain: str, date: str) -> Union[str, bool]:
```
[the remark this answers is not in the document]
Comment (rt17603) [reply openghg/store/_boundary_conditions.py]: Removed `lookup_uuid` method for boundary conditions

File: tests/store/test_boundary_conditions.py (hunk 4 above)
Comment (rt17603) [reply tests/store/test_boundary_conditions.py]: Removed

**Review by gareth-j — APPROVED** [review]
Looks good.
